### ISSUE TITLE
Building custom Ordinal Forest Model in caret's train framework

### ISSUE THREAD

**Issue by HanjoStudy** [topepo/caret#902]
##  Needing help with
I have recently come across the [orindalForest](https://cran.r-project.org/web/packages/ordinalForest/ordinalForest.pdf) package. This package focuses on an implementation of the well known [ranger](https://cran.r-project.org/web/packages/ranger/ranger.pdf) `Random Forest` model. With this in mind, I am trying to integrate the model into the caret framework without any luck. Here is what I have so far:

## The code

Package can be installed from `CRAN`
```{R}
# install.packages("ordinalForest")

# start by naming my method to pass to train
ordinalForest <- list(type = "Classification",
              library = "ordinalForest",
              loop = NULL)

# define the tuning parameters
prm <- data.frame(parameter = c("nsets", "ntreeperdiv", "ntreefinal", "npermtrial", "nbest"),
                  class = rep("numeric", 5),
                  label = c("Score sets", "Number of Trees (small)", "Number of Trees (final)", "Tried score sets" ,"Best score sets" ))

# append them to the list
ordinalForest$parameters <- prm
```
Now that the initial code is written, I can define the default training grid. Some models can do a random search, but I wont implement that as the [paper] (https://epub.ub.uni-muenchen.de/41183/1/TR.pdf) states the defaults are pretty good.
```{R}
ordinalForestGrid <- function(x, y, len = NULL, search = grid) {
  if(search == grid) {
    out <- expand.grid(nsets = 1000,
                       ntreeperdiv = 100,
                       ntreefinal = 5000,
                       npermtrial = 500,
                       nbest = 10)
  } else {
    stop('random search not yet implemented')
  }
  out
}

# append to list
ordinalForest$grid <- ordinalForestGrid
``` 
For the fitting function, the `ordfit` function does have a strange implementation in defining `X` and `Y`. I try and overcome this by explicitly asking for `X` and `Y`, binding them and then implementing the `ordfit` function:
```{R}
ordinalForestFit <- function(x, y) {
  library(ordinalForest)
  
  data <- data.frame(x, Class = y)
  
  ordfor(depvar = "Class", data, nsets = 1000, ntreeperdiv = 100, ntreefinal = 5000,
         perffunction = c("equal"), classimp,
         classweights, nbest = 10, naive = FALSE, num.threads = NULL,
         npermtrial = 500, permperdefault = FALSE)
  
}

# append to list
ordinalForest$fit <- ordinalForestFit
```
I also notice that the classes are stored in a strange slot as well:
```{R}
ordinalForest$levels <- function(x) x@classes
```
Next I add the prediction functions:
```{R}
ordinalForestPred <- function(modelFit, newdata, preProc = NULL, submodels = NULL) {
  predict(modelFit, newdata)
}

# append to list
ordinalForest$predict <- ordinalForestPred

ordinalForestProb <- function(modelFit, newdata, preProc = NULL, submodels = NULL) {
  predict(modelFit, newdata, type = "prob")
}

# append to list
ordinalForest$prob <- ordinalForestProb
```
Laslty I add the sorting, how the tuning parameters are ordered in case similar performance obtained
```{R}
ordinalForestSort <- function (x) x[order(x$nsets, x$ntreeperdiv, x$ntreefinal, 
                                  x$npermtrial, x$nbest), ]

# append to list
ordinalForest$sort <- ordinalForestSort
```
## The test

Now that the hard work is done, lets get the party started

```{R}
# load caret and doParallel library
library(caret)
library(doParallel)

cl <- makePSOCKcluster(3)
registerDoParallel(cl)

# define grid of parameter values
tuneGrid <- expand.grid(nsets = 1000,
                   ntreeperdiv = 100,
                   ntreefinal = 5000,
                   npermtrial = 500,
                   nbest = 10)

set.seed(825)
OFTune <- train(x = hearth[,-11],
                  y = hearth[,11],
                  method = ordinalForest,
                  tuneGrid = tuneGrid, 
                  verbose = TRUE)
Something is wrong; all the Accuracy metric values are missing:
    Accuracy       Kappa    
 Min.   : NA   Min.   : NA  
 1st Qu.: NA   1st Qu.: NA  
 Median : NA   Median : NA  
 Mean   :NaN   Mean   :NaN  
 3rd Qu.: NA   3rd Qu.: NA  
 Max.   : NA   Max.   : NA  
 NA's   :1     NA's   :1    
Error: Stopping
In addition: Warning message:
In nominalTrainWorkflow(x = x, y = y, wts = weights, info = trainInfo,  :
  There were missing values in resampled performance measures.
stopCluster(cl)
```

With great dissapointed, we get the usual fallback error: `There were missing values in resampled performance measures`

**Comment by topepo** [comment]
Can you try this without parallel processing and with a small, reproducible example that I can test with? 

Can you also send the results of `sessionInfo()`?

**Comment by HanjoStudy** [comment]
Updated as requested. A small, reproducible example is available as a gist [here](https://gist.github.com/HanjoStudy/e6d4506c613c1b23bccc86259e558f1d)

**Comment by hadjipantelis** [comment]
@HanjoStudy: Forget `parallel` altogether. The `fit` function is never called. 
The obvious fix for that is to redefine it as: `ordinalForestFit <- function(x, y, param, lev, last, classProbs, ...) { ... }` but then you have other issues as the predict function is also incomplete. 

If you want, I can make a PR about `ordinalForest` during the weekend.

Suggestion for the future: Try and follow the style of other model files. It will save you time and trouble. I am not suggesting that your code structure is worse or the current style is better. Nevertheless, the current style is the existing one, is prevalent through the whole code base and it makes it easy to compare files.

**Comment by HanjoStudy** [comment]
Thanks @hadjipantelis for the advise! I do agree with you, I think next time I would rather start with a caret model skeleton and work it back from there. No rush in implementing the model, I am just slowly trying to wrap my head around the caret framework's underlying components and this `ordinalForest` seemed like a good example to try it out on since it hasn't been implemented.
Ill also keep playing with the code a bit over the next week or so

**Comment by hadjipantelis** [comment]
@topepo Yes, I fully agree (that's why I said it is "incomplete"). OP seems busy so I will drop the PR in the next day or two.

**Comment by hadjipantelis** [comment]
@topepo Some quick questions so we save ourselves back and forth after the PR.
1. I will put this model as "Classification", not "Regression", because the response variable is expected to be a factor. Are you OK wit that?
2. The response variable is not expected to be an "ordered" factor, just a "nominal" factor. Should we enforce that? (Probably this should be addressed at the base package directly but I see a lot of questions about incomprehensible results coming from not enforcing this.)

**Comment by mshad-star** [comment]
Hello.  shouldnt be set hyperparameters in ordinal forests  : nbest = 50 and nset = 10  is not enough and ordinal forest will be suboptimal
